File: 2751-robot-collisions/2751-robot-collisions.go
func survivedRobotsHealths(positions []int, healths []int, directions string) []int {
	n := len(positions)
	robots := []Robot{}
	for i := 0; i < n; i++ {
		robots = append(robots, Robot{
			num:    i + 1,
			pos:    positions[i],
			health: healths[i],
			dir:    directions[i],
		})
	}

    // sort by pos
	sort.Slice(robots, func(i, j int) bool { return robots[i].pos < robots[j].pos })

	stack := []Robot{}
	for _, robot := range robots {
		for len(stack) > 0 && stack[len(stack)-1].dir == 'R' && robot.dir == 'L' && stack[len(stack)-1].health < robot.health {
			robot.health--
			stack = stack[:len(stack)-1]
		}

		if len(stack) > 0 && stack[len(stack)-1].dir == 'R' && robot.dir == 'L' && stack[len(stack)-1].health > robot.health {
			stack[len(stack)-1].health--
		} else if len(stack) > 0 && stack[len(stack)-1].dir == 'R' && robot.dir == 'L' && stack[len(stack)-1].health == robot.health {
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, robot)
		}
	}

    // sort by robot num
    sort.Slice(stack, func(i, j int) bool { return stack[i].num < stack[j].num })

    ans := []int{}
    for _, s := range stack {
        ans = append(ans, s.health)
    }

    return ans
}

type Robot struct {
	num    int
	pos    int
	health int
	dir    byte
}

// array, sorting, stack
// time: O(nlog(n))
// space: O(n)

File: 2751-robot-collisions/2751-robot-collisions.py
class Solution:
    def survivedRobotsHealths(self, positions: List[int], healths: List[int], directions: str) -> List[int]:
        n = len(positions)

        # (robot's original idx, pos, health, dir)
        robots = [(i + 1, positions[i], healths[i], directions[i]) for i in range(n)]

        robots.sort(key=lambda x: x[1])

        stack = []
        for idx, pos, health, dir in robots:
            curr_health = health

            while stack and stack[-1][3] == 'R' and dir == 'L' and stack[-1][2] < curr_health:
                curr_health -= 1
                stack.pop()
            
            if stack and stack[-1][3] == 'R' and dir == 'L' and stack[-1][2] > curr_health:
                top = stack[-1]
                stack[-1] = (top[0], top[1], top[2] - 1, top[3])
            elif stack and stack[-1][3] == 'R' and dir == 'L' and stack[-1][2] == curr_health:
                stack.pop()
            else:
                stack.append((idx, pos, curr_health, dir))
        
        stack.sort(key=lambda x: x[0])

        return [robot[2] for robot in stack]

# array, sorting, stack
# time: O(nlog(n))
# space: O(n)